source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1551,18 +1552,28 @@ ClusterManagerImpl::allocateOdCdsApi(OdCdsCreationFunction creation_function,
                                      const envoy::config::core::v3::ConfigSource& odcds_config,
                                      OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
                                      ProtobufMessage::ValidationVisitor& validation_visitor) {
-  // TODO(krnowak): Instead of creating a new handle every time, store the handles internally and
-  // return an already existing one if the config or locator matches. Note that this may need a
-  // way to clean up the unused handles, so we can close the unnecessary connections.
+  // Generate a unique key for this subscription based on config and locator.
+  // This enables reuse of subscriptions with the same configuration.
+  uint64_t config_hash = MessageUtil::hash(odcds_config);
+  if (odcds_resources_locator.has_value()) {
+    config_hash = absl::HashOf(config_hash, MessageUtil::hash(*odcds_resources_locator));
+  }
+
+  auto it = odcds_subscriptions_.find(config_hash);
+  if (it != odcds_subscriptions_.end()) {
+    return OdCdsApiHandleImpl::create(*this, config_hash);
+  }
+
   auto odcds_or_error =
       creation_function(odcds_config, odcds_resources_locator, xds_manager_, *this, *this,
                         *stats_.rootScope(), validation_visitor, context_);
   RETURN_IF_NOT_OK_REF(odcds_or_error.status());
-  return OdCdsApiHandleImpl::create(*this, std::move(*odcds_or_error));
+  odcds_subscriptions_.emplace(config_hash, std::move(*odcds_or_error));
```

Comment:
Can you elaborate on the lifetime of the subscription?
Specifically I see that elements are inserted into `odcds_subscriptions_`, but I don't see where they are removed.